### PR TITLE
feat(cdk): support encoding Tokens as numbers

### LINKED ISSUE #1455: Tokens: be able to represent as numbers
We are able to represent tokens as strings (`token.toString()`) and as string arrays (`token.toList()`), but in certain cases, there's a need to represent them as numbers. Introduce `token.toNumber()`.

We need to somehow encode the fact that this is a token into a number. It's not going to be elegant (e.g. `-66737346654.<random>` can be the key to the token map).

Related to #744 
Related to #1453

### PR DESCRIPTION
**Pull request by rix0rrr**
Numbers can now be encoded into a set of (hugely negative) numbers,
and the encoding can be reversed.

This allows APIs that take numbers to take lazy values and intrinsics
that evaluate to numbers.

This change only introduces the capability, it does not use it in any
of the construct libraries yet.

Fixes #1455.

----

### Pull Request Checklist

* [ ] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [ ] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
  - __Design__: For significant features, design document added to `design` folder
* [ ] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** and module name in parens, which will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [ ] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

**Comment by RomainMuller** [comment]
Is it possible to isolate the `package-lock.json` changes to a separate PR? I feel like we should better have specific reverse-ability for this PR (although I haven't read the code just yet, so maybe I'm over-indexing on the implications).

File: packages/@aws-cdk/cdk/lib/encoding.ts [hunk 1]
```diff
@@ -200,9 +200,79 @@ export function containsListTokenElement(xs: any[]) {
...
+ * significant exponent bit.
+ *
+ * Currently not supporting BE architectures.
+ */
+// tslint:disable-next-line:no-bitwise
+const DOUBLE_TOKEN_MARKER_BITS = 0xFBFF << 16;
+
+/**
+ * Return a special Double value that encodes the given integer
+ */
```
Comment (eladb) [inline packages/@aws-cdk/cdk/lib/encoding.ts]: Mention in the docstring that this is used to encode an _index_ to the token map and therefore it's okay that we don't support negative numbers. Also, mention the limit on the number of number tokens we can support.

File: packages/@aws-cdk/cdk/lib/encoding.ts [hunk 2]
```diff
@@ -200,9 +200,79 @@ export function containsListTokenElement(xs: any[]) {
...
+ * into octets, then grouped into 32-bit words):
+ *
+ * mmmmmmm.mmmmmmm.mmmmmmm.mmmmmmm | mmmmmmm.mmmmmmm.EEEEEmm.sEEEEEE
+ *
+ * - m: mantissa (52 bits)
+ * - E: exponent (11 bits)
+ * - s: sign (1 bit)
+ *
+ * We put the following marker into the top 16 bits (exponent and sign), and
+ * use the mantissa part to encode the token index. To save some bit twiddling
```
Comment (eladb) [inline packages/@aws-cdk/cdk/lib/encoding.ts]: "encode the index to the token in the token map"

File: packages/@aws-cdk/cdk/lib/token-map.ts [hunk 3]
```diff
@@ -49,6 +52,15 @@ export class TokenMap {
     return [`${BEGIN_LIST_TOKEN_MARKER}${key}${END_TOKEN_MARKER}`];
   }
 
+  /**
+   * Create a unique number representation for this Token and return it
+   */
+  public registerNumber(token: Token): number {
+    const tokenIndex = this.tokenCounter++;
```
Comment (eladb) [inline packages/@aws-cdk/cdk/lib/token-map.ts]: Error if we reached the limit.

**Pull request merged**